pytest/lib/mocknet.py

```diff
@@ -0,0 +1,188 @@
+KEY_TARGET_ENV_VAR = 'NEAR_PYTEST_KEY_TARGET'
+DEFAULT_KEY_TARGET = '/tmp/mocknet'
+
+TMUX_STOP_SCRIPT = '''
+while tmux has-session -t near; do
+tmux kill-session -t near || true
+done
+'''
+
+TMUX_START_SCRIPT = '''
+sudo rm -rf /tmp/near.log
+tmux new -s near -d bash
+tmux send-keys -t near 'RUST_BACKTRACE=full /home/ubuntu/near run 2>&1 | tee /home/ubuntu/near.log' C-m
+'''
+
+
+def get_nodes():
+    nodes = [GCloudNode(f'{NODE_BASE_NAME}{i}') for i in range(0, NUM_NODES)]
+    for n in nodes:
+        n.machine.username = NODE_USERNAME
+        n.machine.ssh_key_path = NODE_SSH_KEY_PATH
+    return nodes
+
+
+def create_target_dir(machine):
+    base_target_dir = os.environ.get(KEY_TARGET_ENV_VAR, DEFAULT_KEY_TARGET)
+    target_dir = f'{base_target_dir}/{machine.name}'
+    run(f'mkdir -p {target_dir}')
+    return target_dir
+
+
+def get_validator_account(node):
+    m = node.machine
+    target_dir = create_target_dir(m)
+    m.download(f'/home/ubuntu/.near/validator_key.json', target_dir)
+    return Key.from_json_file(f'{target_dir}/validator_key.json')
+
+
+def get_epoch_length_in_blocks(node):
+    m = node.machine
+    target_dir = create_target_dir(m)
+    m.download(f'/home/ubuntu/.near/genesis.json', target_dir)
+    with open(f'{target_dir}/genesis.json') as f:
+        config = json.load(f)
+        epoch_length_in_blocks = config['epoch_length']
+        return epoch_length_in_blocks
+
+
+# Sends the transaction to the network via `node` and checks for success.
+# Some retrying is done when the node returns a Timeout error.
+def send_transaction(node, tx, tx_hash, account_id, timeout=30):
+    response = node.send_tx_and_wait(tx, timeout)
+    loop_start = time.time()
+    missing_count = 0
+    while 'error' in response.keys():
+        error_data = response['error']['data']
+        if error_data == 'Timeout':
+            print(
+                f'WARN: transaction {tx_hash} returned Timout, checking status again.'
+            )
+            time.sleep(2)
+            response = node.get_tx(tx_hash, account_id)
+        elif "doesn't exist" in error_data:
+            missing_count += 1
+            print(
+                f'WARN: transaction {tx_hash} falied to be recieved by the node, checking again.'
+            )
+            if missing_count < 10:
+                time.sleep(2)
+                response = node.get_tx(tx_hash, account_id)
+            else:
+                print(f'WARN: re-sending transaction {tx_hash}.')
+                response = node.send_tx_and_wait(tx, timeout)
+                missing_count = 0
+        else:
+            raise RuntimeError(
+                f'Error in processing transaction {tx_hash}: {response}')
+        if time.time() - loop_start > timeout:
+            raise TimeoutError(
+                f'Transaction {tx_hash} did not complete successfully within the timeout'
+            )
+
+    if 'SuccessValue' not in response['result']['status'].keys():
```

Comment:
`.keys()` is not necessary

Reply:
Done in https://github.com/nearprotocol/nearcore/pull/2868/commits/ab35f41470c40ebf8881caf85f9cd4ae477b52de